src/stash/src/postgres.rs

```diff
@@ -88,13 +105,36 @@ impl Postgres {
         if fence_oid.is_none() {
             tx.batch_execute(SCHEMA).await?;
         }
-        // Bump the epoch, which will cause any previous connection to fail.
+        // Bump the epoch, which will cause any previous connection to fail. Add a
+        // unique rng so that if some other thing recreates the entire schema, we
+        // can't accidentally have the same epoch, rng pair (especially risky if the
+        // current epoch has been bumped exactly once, then gets recreated by another
+        // connection that also bumps it once).
         let epoch = tx
-            .query_one("UPDATE fence SET epoch=epoch+1 RETURNING epoch", &[])
+            .query_one(
+                "UPDATE fence SET epoch=epoch+1, rng=$1 RETURNING epoch",
+                &[&conn.rng.to_vec()],
+            )
             .await?
             .get(0);
         tx.commit().await?;
-        Ok(Postgres { conn, epoch })
+        conn.epoch = epoch;
+        Ok(conn)
+    }
+
+    /// Creates a new connection.
+    async fn connect(&mut self) -> Result<(), StashError> {
+        let (client, connection) = tokio_postgres::connect(&self.url, NoTls).await?;
+        mz_ore::task::spawn(|| "tokio-postgres stash connection", async move {
+            if let Err(e) = connection.await {
+                eprintln!("postgres stash connection error: {}", e);
```

Comment:
```suggestion
                tracing::error!("postgres stash connection error: {}", e);
```

is the idea that the `client` will fail, which causes this method to be called again, which spawns a new connection task?

Reply:
Yes, in `transact_inner`, the client is reconnected if it has closed.